ldif.go

```diff
@@ -0,0 +1,256 @@
+// records are not supported.
+type LDIF struct {
+	RelaxedParser bool // ignore parsing errors in a line
+	Entries       []*Entry
+	curLine       int
+	version       int
+	changeType    string
+}
+
+var CR byte = '\x0D'
+var LF byte = '\x0A'
+var SEP string = string([]byte{CR, LF})
+var Comment byte = '#'
+var SPACE byte = ' '
+var SPACES = string(SPACE)
+
+func (l *LDIF) newError(msg string) error {
+	return errors.New(fmt.Sprintf("error on line %d: %s\n", l.curLine, msg))
+}
+
+// Parses the LDIF, the caller is responsible for closing the io.Reader if that's
+// needed. This Parser may be called several times with different io.Readers to
+// combine the files.
+func (l *LDIF) Parse(r io.Reader) (err error) {
+	if r == nil {
+		return errors.New("No reader present")
+	}
+	reader := bufio.NewReader(r)
+	l.curLine = 0
+	l.changeType = ""
+	l.version = 0
+
+	var lines [][]byte
+	var line, nextLine []byte
+
+	for {
+		l.curLine++
+		nextLine, err = reader.ReadBytes(LF)
+		nextLine = bytes.TrimRight(nextLine, SEP)
+		// fmt.Fprintf(os.Stderr, "NEXT=>%s<\n", nextLine)
+		switch err {
+		case nil, io.EOF:
+			switch len(nextLine) {
+			case 0:
+				if len(line) == 0 && err == io.EOF {
+					return nil
+				}
+				lines = append(lines, line)
+				entry, perr := l.parseEntry(lines)
+				if perr != nil {
+					if !l.RelaxedParser {
+						return l.newError(perr.Error())
+					}
+				}
+				l.Entries = append(l.Entries, entry)
+				line = []byte{}
+				lines = [][]byte{}
+				if err == io.EOF {
+					return nil
+				}
+			default:
+				switch nextLine[0] {
+				case Comment:
+					continue
+				case SPACE:
+					if len(nextLine) > 1 {
+						line = append(line, nextLine[1:]...)
+						continue
+					} else {
+						return l.newError("space only line")
+					}
+				default:
+					if len(line) != 0 {
+						lines = append(lines, line)
+					}
+					line = nextLine
+					continue
+				}
+			}
+		default:
+			return l.newError(err.Error())
+		}
+	}
+	if len(lines) != 0 {
+		entry, perr := l.parseEntry(lines)
+		if perr != nil {
+			return l.newError(perr.Error())
+		}
+		l.Entries = append(l.Entries, entry)
+	}
+	return nil
+}
+
+func (l *LDIF) parseEntry(lines [][]byte) (entry *Entry, err error) {
+	// for i, line := range lines {
+	//     fmt.Fprintf(os.Stderr, "% 2d %s\n", i, line)
+	// }
+	if l.version == 0 && bytes.HasPrefix(lines[0], []byte("version:")) {
+		line := bytes.TrimLeft(lines[0][8:], SPACES)
+		if version, err := strconv.Atoi(string(line)); err != nil {
+			return nil, err
+		} else {
+			if version != 1 {
+				return nil, errors.New("Invalid version spec " + string(line))
+			}
+			l.version = 1
+			lines = lines[1:]
+		}
+	}
+
+	if !bytes.HasPrefix(lines[0], []byte("dn:")) {
+		return nil, errors.New("Missing dn:")
+	}
+	_, val, err := l.parseLine(lines[0])
+	if err != nil {
+		return nil, err
+	}
+	dn := val
+
+	lines = lines[1:]
+	if bytes.HasPrefix(lines[0], []byte("changetype:")) {
```

Comment:
seems like there are several chances to panic with accesses to `lines[0]` if too few lines are passed in